Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt

```diff
@@ -0,0 +1,93 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.paintroid.tools.helper
+
+import android.graphics.Bitmap
+import android.graphics.Color
+import android.graphics.Rect
+import android.util.Log
+
+private const val TAG = "cropAlgorithmSnail"
+
+class JavaCropAlgorithm : CropAlgorithm {
+    private fun getRange(s: Int, e: Int): ArrayList<Int> {
+        val range = ArrayList<Int>()
+        for (i in s..e) range.add(i)
+        return range
+    }
+
+    private fun getTopmostCellWithOpaquePixel(bounds: Rect, pixels: Array<IntArray>): Int? =
+        getRange(bounds.top, bounds.bottom).firstOrNull { row ->
+            getRange(
+                bounds.left,
+                bounds.right
+            ).firstOrNull { col -> pixels[row][col] != Color.TRANSPARENT } != null
+        }
+
+    private fun getBottommostCellWithOpaquePixel(bounds: Rect, pixels: Array<IntArray>): Int? =
+        getRange(bounds.top, bounds.bottom).lastOrNull { row ->
+            getRange(
+                bounds.left,
+                bounds.right
+            ).firstOrNull { col -> pixels[row][col] != Color.TRANSPARENT } != null
+        }
+
+    private fun getLeftmostCellWithOpaquePixel(bounds: Rect, pixels: Array<IntArray>): Int? =
+        getRange(bounds.left, bounds.right).firstOrNull { col ->
+            getRange(
+                bounds.top,
+                bounds.bottom
+            ).firstOrNull { row -> pixels[row][col] != Color.TRANSPARENT } != null
+        }
+
+    private fun getRightmostCellWithOpaquePixel(bounds: Rect, pixels: Array<IntArray>): Int? =
+        getRange(bounds.left, bounds.right).lastOrNull { col ->
+            getRange(
+                bounds.top,
+                bounds.bottom
+            ).firstOrNull { row -> pixels[row][col] != Color.TRANSPARENT } != null
+        }
+
+    private fun initPixels(bitmap: Bitmap): Array<IntArray> {
+        val pixels = Array(bitmap.height) { IntArray(bitmap.width) }
+        for (i in 0 until bitmap.height) {
+            bitmap.getPixels(pixels[i], 0, bitmap.width, 0, i, bitmap.width, 1)
+        }
+        return pixels
+    }
+
+    override fun crop(bitmap: Bitmap?): Rect? {
+        if (bitmap == null) {
+            Log.e(TAG, "bitmap is null!")
+            return null
+        }
+        val pixels = initPixels(bitmap)
+        val bounds = Rect(0, 0, bitmap.width - 1, bitmap.height - 1)
+        val top = getTopmostCellWithOpaquePixel(bounds, pixels)
+        if (top == null) {
+            Log.i(TAG, "nothing to crop")
+            return top
+        }
+        bounds.top = top
```

Comment:
```suggestion
        bounds.top = getTopmostCellWithOpaquePixel(bounds, pixels) ?: run { 
            Log.i(TAG, "nothing to crop") 
            return null
        }
```

Reply:
Prev code bit obfuscating, since top is a nullable integer which was returned for a nullable Rect.

Reply:
Since this seems more like a debug log, you could simply delete it.
```suggestion
        bounds.top = getTopmostCellWithOpaquePixel(bounds, pixels) ?: return null
```

Reply:
deleted the debug log.